pkg/master/controller/clusterauthenticationtrust/cluster_authentication_trust_controller.go

```diff
@@ -0,0 +1,504 @@
+	ret := ClusterAuthenticationInfo{}
+
+	var err error
+	ret.RequestHeaderGroupHeaders, err = jsonDeserializeStringSlice(data["requestheader-group-headers"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderExtraHeaderPrefixes, err = jsonDeserializeStringSlice(data["requestheader-extra-headers-prefix"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderAllowedNames, err = jsonDeserializeStringSlice(data["requestheader-allowed-names"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+	ret.RequestHeaderUsernameHeaders, err = jsonDeserializeStringSlice(data["requestheader-username-headers"])
+	if err != nil {
+		return ClusterAuthenticationInfo{}, err
+	}
+
+	if caBundle := data["requestheader-client-ca-file"]; len(caBundle) > 0 {
+		ret.RequestHeaderCA, err = dynamiccertificates.NewStaticCAContent("existing", []byte(caBundle))
+		if err != nil {
+			return ClusterAuthenticationInfo{}, err
+		}
+	}
+
+	if caBundle := data["client-ca-file"]; len(caBundle) > 0 {
+		ret.ClientCA, err = dynamiccertificates.NewStaticCAContent("existing", []byte(caBundle))
+		if err != nil {
+			return ClusterAuthenticationInfo{}, err
+		}
+	}
+
+	return ret, nil
+}
+
+func jsonSerializeStringSlice(in []string) (string, error) {
+	out, err := json.Marshal(in)
+	if err != nil {
+		return "", err
+	}
+	return string(out), err
+}
+
+func jsonDeserializeStringSlice(in string) (headerrequest.StringSliceProvider, error) {
+	if len(in) == 0 {
+		return nil, nil
+	}
+
+	out := []string{}
+	if err := json.Unmarshal([]byte(in), &out); err != nil {
+		return nil, err
+	}
+	return headerrequest.StaticStringSlice(out), nil
+}
+
+func combineStringSlices(lhs, rhs headerrequest.StringSliceProvider) headerrequest.StringSliceProvider {
+	ret := []string{}
+	present := sets.String{}
+
+	if lhs != nil {
+		for _, curr := range lhs.Value() {
+			if present.Has(curr) {
+				continue
+			}
+			ret = append(ret, curr)
+			present.Insert(curr)
+		}
+	}
+
+	if rhs != nil {
+		for _, curr := range rhs.Value() {
+			if present.Has(curr) {
+				continue
+			}
+			ret = append(ret, curr)
+			present.Insert(curr)
+		}
+	}
+
+	return headerrequest.StaticStringSlice(ret)
+}
+
+func combineCertLists(lhs, rhs dynamiccertificates.CAContentProvider) (dynamiccertificates.CAContentProvider, error) {
+	certificates := []*x509.Certificate{}
+
+	if lhs != nil {
+		lhsCABytes := lhs.CurrentCABundleContent()
+		lhsCAs, err := cert.ParseCertsPEM(lhsCABytes)
+		if err != nil {
+			return nil, err
+		}
+		certificates = append(certificates, lhsCAs...)
+	}
+	if rhs != nil {
+		rhsCABytes := rhs.CurrentCABundleContent()
+		rhsCAs, err := cert.ParseCertsPEM(rhsCABytes)
+		if err != nil {
+			return nil, err
+		}
+		certificates = append(certificates, rhsCAs...)
+	}
+
+	certificates = filterExpiredCerts(certificates...)
+
+	finalCertificates := []*x509.Certificate{}
+	// now check for duplicates. n^2, but super simple
```

Comment:
~~one of lhs or rhs is small => n^2 is fine~~

Reply:
isn't there some unique property in the CA you can use instead?

How big can n get? Another cert per rotation? Rotation is each 30 days => 12 per year. So probably not so big.

Reply:
> isn't there some unique property in the CA you can use instead?
> 
> How big can n get? Another cert per rotation? Rotation is each 30 days => 12 per year. So probably not so big.

You're looking at order of 10 infrequently checked.  I think this is fine until you see it in a profile.